src/rust.rs

```diff
@@ -270,41 +268,103 @@ impl RootKind for Value {
     fn rootKind() -> jsapi::RootKind { jsapi::RootKind::Value }
 }
 
-impl<T: RootKind + Copy> Rooted<T> {
-    pub fn new_with_addr(cx: *mut JSContext, initial: T, addr: *const u8) -> Rooted<T> {
-        let ctxfriend: &mut ContextFriendFields = unsafe {
-            mem::transmute(cx)
-        };
-
-        let kind = T::rootKind() as usize;
-        let root = Rooted::<T> {
+impl<T> Rooted<T> {
+    pub fn new_unrooted(initial: T) -> Rooted<T> {
+        Rooted {
             _base: RootedBase { _phantom0: PhantomData },
-            stack: &mut ctxfriend.roots.stackRoots_[kind] as *mut _ as *mut _,
-            prev: ctxfriend.roots.stackRoots_[kind] as *mut _,
+            stack: ptr::null_mut(),
+            prev: ptr::null_mut(),
             ptr: initial,
-        };
+        }
+    }
+
+    pub unsafe fn add_to_root_stack(&mut self, cx: *mut JSContext) where T: RootKind {
+        let ctxfriend: &mut ContextFriendFields = mem::transmute(cx);
+
+        let kind = T::rootKind() as usize;
+        self.stack = &mut ctxfriend.roots.stackRoots_[kind] as *mut _ as *mut _;
+        self.prev = ctxfriend.roots.stackRoots_[kind] as *mut _;
 
-        ctxfriend.roots.stackRoots_[kind] = unsafe { mem::transmute(addr) };
-        root
+        ctxfriend.roots.stackRoots_[kind] = self as *mut _ as usize as _;
     }
 
-    pub fn new(cx: *mut JSContext, initial: T) -> Rooted<T> {
-        Rooted::new_with_addr(cx, initial, unsafe { return_address() })
+    pub unsafe fn remove_from_root_stack(&mut self) {
+        assert!(*self.stack == mem::transmute(&*self));
+        *self.stack = self.prev;
     }
+}
 
-    pub fn handle(&self) -> Handle<T> {
+/// Rust API for keeping a Rooted value in the context's root stack.
+/// Example usage: `rooted!(in(cx) let x = UndefinedValue());`.
+/// `RootedGuard::new` also works, but the macro is preferred.
+pub struct RootedGuard<'a, T: 'a> {
+    root: &'a mut Rooted<T>
```

Comment:
The idea is that the `Rooted<T>` can safely be added to the root stack because it cannot be moved by the compiler while there is an active `&mut` reference to it?